### ISSUE TITLE
`aria-allowed-role` issues with axe-core>=3.1.0  (PR #1118)

### ISSUE THREAD

**Issue by majapw** [dequelabs/axe-core#1110]
<pre><code>axe-core version: 3.1.0</code></pre>

Hi there! (Sorry for opening so many issues)

Since upgrading our internal version, I've also noticed some issues with other tests and the new `allowed-role` rule. 

Specifically we have an input with `type="text"` and `role="textbox"` failing the `aria-allowed-role` spec:
```
Message: ARIA role must be appropriate for the element
Read more: https://dequeuniversity.com/rules/axe/3.1/aria-allowed-role?application=axeAPI
Code: <input autocomplete="off" id="input_1" name="input_1" placeholder="This is placeholder text!" role="textbox" value="Re" type="text">
```

Reading through https://www.w3.org/TR/html5/dom.html#sec-strong-native-semantics, this seems like it should be allowed by this rule (although perhaps it is failing because the role is redundant---if that is the case, maybe the message should be different?).

**Comment by jkodu** [comment]
@majapw 

Thanks again. No need to apologise for helping us with valid feedback. We appreciate it.

Once again, let us run a few tests on this and get back. I agree that, the message could be more helpful. Once we narrow down the reason, will circle back with information. The code snippet is very helpful. 

Keep it coming.

**Comment by majapw** [comment]
Thanks @jkodu I did some more testing, and it does seem like this issue stems from the redundant role, rather than the role being inappropriate. It would be great if there was some differentiation between these two scenarios as the messaging is pretty ambiguous. 

Basically, it would be great to differentiate between the two cases:
```
<input type="text" role="foobar" /> // role is inappropriate for the element
```
and
```
<input type="text" role="textbox" /> // role is redundant for the element
```

I think the messaging is otherwise very confusing! :) (but also shoutout to `eslint-plugin-jsx-a11y` for helping me find the underlying cause)

**Comment by WilcoFiers** [comment]
Neither of those cases should be failing. `role=foobar` is invalid and should be ignored by aria-allowed-role, and redundant roles should be permitted by default. There's an option in the `aria-allowed-role` check to not allow it, but it should be allowed by default.

**Comment by jkodu** [comment]
PR - https://github.com/dequelabs/axe-core/pull/1118

**Comment by majapw** [comment]
I am also seeing the same error on the following code:
```
Message: ARIA role must be appropriate for the element
--
Read more: https://dequeuniversity.com/rules/axe/3.1/aria-allowed-role?application=axeAPI
Code: <button style="..." role="gridcell" title="IconCheckmark" aria-label="IconCheckmark icon">
```
and I'm not entirely sure why. The button is wrapped in a 
```
<div role="grid">
  <div role="row">
    ...
  </div>
</div>
```
This same structure is used in a different test as well where it does not seem to error and that is somewhat confusing as well.

**Comment by jkodu** [comment]
Hi @majapw 

With regards to the code snippet you provided above, I believe this is indeed a violation.

From specs:
![image](https://user-images.githubusercontent.com/20978252/45083529-f3492a00-b0f3-11e8-907b-04692194f09e.png)


Further reference:
https://www.w3.org/TR/html-aria/

Although, `button` is nested with in `grid` and `row`. The `button` element only allows certain roles and is not open to `any` role. In this case, one of the solutions could be to wrap the `button` with a `span` perhaps with the role of `gridcell`.

**Comment by jkodu** [comment]
PR has been merged. Closing issue.
